webserver/templates/user/scraper.js

```diff
@@ -165,77 +175,76 @@ function reportScrobbles(struct) {
             times4Error++;
             // We mark 4xx errors as completed because we don't
             // retry them
-            numCompleted++;
             console.log("4xx error, skipping");
+            pageDone();
         } else if (this.status >= 500) {
             console.log("received http error " + this.status + " req'ing");
             times5Error++;
-            enqueueReport(struct);
+            reportScrobbles(struct);
+        } else {
+            console.log("received http status " + this.status + ", skipping");
+            pageDone();
         }
         if (numCompleted >= numberOfPages) {
             updateMessage("<i class='fa fa-check'></i> Import finished<br><span style='font-size:8pt'>Thank you for using ListenBrainz</span>");
         } else {
             updateMessage("<i class='fa fa-cog fa-spin'></i> Sending page " + numCompleted + " of " + numberOfPages + " to ListenBrainz<br><span style='font-size:8pt'>Please don't navigate while this is running</span>");
         }
-        getNextPageIfSlots();
     };
     xhr.ontimeout = function(context) {
         console.log("timeout, req'ing");
-        enqueueReport(struct);
+        reportScrobbles(struct);
     }
     xhr.onabort = function(context) {
         console.log("abort, req'ing");
-        enqueueReport(struct);
+        reportScrobbles(struct);
     };
     xhr.onerror = function(context) {
         console.log("error, req'ing");
-        enqueueReport(struct);
+        reportScrobbles(struct);
     };
-    xhr.onloadend = function(context) {
-        activeSubmissions--;
-    }
     xhr.send(JSON.stringify(struct));
 }
 
-function reportPage(response) {
-    var elem = document.createElement("div");
-    elem.innerHTML = response;
-    var struct = encodeScrobbles(elem);
-    enqueueReport(struct);
-}
-
 function reportPageAndGetNext(response) {
     timesGetPage++;
     if (page == 1) {
-      updateMessage("<i class='fa fa-cog fa-spin'></i> working<br><span style='font-size:8pt'>Please don't navigate away from this page while the process is running</span>");
+        updateMessage("<i class='fa fa-cog fa-spin'></i> working<br><span style='font-size:8pt'>Please don't navigate away from this page while the process is running</span>");
     }
-    reportPage(response);
-
-    getNextPageIfSlots();
+    var elem = document.createElement("div");
+    elem.innerHTML = response;
+    var struct = encodeScrobbles(elem);
+    reportScrobbles(struct);
 }
 
-function getNextPageIfSlots() {
+function getNextPagesIfSlots() {
     // Get a new lastfm page and queue it only if there are more pages to download and we have
     // less than 10 pages waiting to submit
-    if (page <= numberOfPages && activeSubmissions < 10) {
+    while (page <= numberOfPages && activeSubmissions < 10) {
         page += 1;
-        setTimeout(function() { getLastFMPage(page, reportPageAndGetNext) }, 0 + Math.random()*100);
```

Comment:
This was here because originally we didn't have a limit on the number of active submission. This was a random pause between paging between last.fm pages.   I wasn't really worried about having 11 or 12 activeSubmissions going, just as long as it was about 10 instead of about 100